Sources/Hummingbird/Utils/ByteBuffer+foundation.swift

```diff
@@ -0,0 +1,172 @@
+        let index = index0 - self.readerIndex
+        guard index >= 0 && length >= 0 && index <= self.readableBytes - length else {
+            return nil
+        }
+        let doCopy: Bool
+        switch byteTransferStrategy {
+        case .copy:
+            doCopy = true
+        case .noCopy:
+            doCopy = false
+        case .automatic:
+            doCopy = length <= 256 * 1024
+        }
+
+        return self.withUnsafeReadableBytesWithStorageManagement { ptr, storageRef in
+            if doCopy {
+                return Data(
+                    bytes: UnsafeMutableRawPointer(mutating: ptr.baseAddress!.advanced(by: index)),
+                    count: Int(length)
+                )
+            } else {
+                _ = storageRef.retain()
+                return Data(
+                    bytesNoCopy: UnsafeMutableRawPointer(mutating: ptr.baseAddress!.advanced(by: index)),
+                    count: Int(length),
+                    deallocator: .custom { _, _ in storageRef.release() }
+                )
+            }
+        }
+    }
+
+    /// Read `length` bytes off this `ByteBuffer`, move the reader index forward by `length` bytes and return the result
+    /// as `Data`.
+    ///
+    /// - parameters:
+    ///     - length: The number of bytes to be read from this `ByteBuffer`.
+    ///     - byteTransferStrategy: Controls how to transfer the bytes. See `ByteTransferStrategy` for an explanation
+    ///                             of the options.
+    /// - returns: A `Data` value containing `length` bytes or `nil` if there aren't at least `length` bytes readable.
+    package mutating func _readData(length: Int, byteTransferStrategy: _ByteTransferStrategy) -> Data? {
+        guard
+            let result = self._getData(at: self.readerIndex, length: length, byteTransferStrategy: byteTransferStrategy)
+        else {
+            return nil
+        }
+        self.moveReaderIndex(forwardBy: length)
+        return result
+    }
+
+    /// Attempts to decode the `length` bytes from `index` using the `JSONDecoder` `decoder` as `T`.
+    ///
+    /// - parameters:
+    ///    - type: The type type that is attempted to be decoded.
+    ///    - decoder: The `JSONDecoder` that is used for the decoding.
+    ///    - index: The index of the first byte to decode.
+    ///    - length: The number of bytes to decode.
+    /// - returns: The decoded value if successful or `nil` if there are not enough readable bytes available.
+    @inlinable
+    package func _getJSONDecodable<T: Decodable>(
+        _ type: T.Type,
+        decoder: JSONDecoder = JSONDecoder(),
+        at index: Int,
+        length: Int
+    ) throws -> T? {
+        guard let data = self._getData(at: index, length: length, byteTransferStrategy: .noCopy) else {
+            return nil
+        }
+        return try decoder.decode(T.self, from: data)
```

Comment:
Can we do a `withData`-style API instead? The `noCopy` byteTransferStrategy seems like a bad idea to me, since it's easily miswielded.

Reply:
These functions are direct copies of the functions in NIOFoundationCompat, except I added an `_` in front of them. The `ByteTransferStrategy.noCopy` works as far as I know. NIO seemed happy with it.

Reply:
That's fair, I don't think we need it here - but I agree we shouldn't diverse a "fork"